File: 1036-rotting-oranges/1036-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        if(grid == null || grid.length == 0){
            return -1;
        }

        Queue<int[]> q = new LinkedList<>();
        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j < grid[0].length; j++){
                if(grid[i][j] == 2){
                    q.offer(new int[]{i,j});
                }
            }
        }

        int cnt = countRotten(grid,q);
        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j < grid[0].length; j++){
                if(grid[i][j] == 1) return -1;
            }
        }
 
        return (cnt == 0) ? 0 : cnt-1;
    }
    public int countRotten(int[][] grid, Queue<int[]> q){
        int m = grid.length;
        int n = grid[0].length;
        int[][] dirs = {{1,0},{-1,0},{0,1},{0,-1}}; 
        int cnt  = 0;

        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                int[] cell = q.poll(); 
                for(int[] dir: dirs){
                    int newR = cell[0] + dir[0];
                    int newC = cell[1] + dir[1];

                    if(newR >= 0 && newR < m && newC >= 0 && newC < n && grid[newR][newC] == 1){
                        grid[newR][newC] = 2;
                        q.offer(new int[]{newR,newC});
                    }
                }
            }
            cnt++;
        }

        return cnt;
    }
}

File: 1036-rotting-oranges/1036-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid):
        if not grid:
            return -1

        q = deque()
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 2:
                    q.append((i, j))

        cnt = self.countRotten(grid, q)

        for row in grid:
            if 1 in row:
                return -1

        return 0 if cnt == 0 else cnt - 1

    def countRotten(self, grid, q):
        m, n = len(grid), len(grid[0])
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        cnt = 0

        while q:
            for _ in range(len(q)):
                r, c = q.popleft()
                for dr, dc in directions:
                    new_r, new_c = r + dr, c + dc

                    if 0 <= new_r < m and 0 <= new_c < n and grid[new_r][new_c] == 1:
                        grid[new_r][new_c] = 2
                        q.append((new_r, new_c))

            cnt += 1

        return cnt
